Analyzer: Consolidate some plugin meta-functions

tcoyvwac opened the pull request:
Changes: 
* Prefer using standard algorithms, to help check if plug-In is available.
* Extract some common code between the two Analyzer{Beats,Keys} classes into its own meta-interface class.

---

_Notes:_
* _I am flexible in adjusting this PR so only the first two commits (which add the standard algorithm code) are accepted._
*  _If not allowed (to adjust the PR), then I am happy to make another PR just for adding the standard algorithm code alone (if this PR is not merged). Thanks!_ :smile:

tcoyvwac committed:
analyzer: prefer standard algorithms

analyzer{beats,key}.cpp:
Prefer std::any_of over ranged-for.

tcoyvwac committed:
analyzer: make static function: matchAndSetPluginId()

Promoted common closure code into a static member function.

tcoyvwac committed:
analyzer: added analyzerpluginsupport.{cpp,h}

analyzerbeats.h:
- Extracted common AnalyzerPluginSupportInfo_{A,B} interface to own
header.
- Renamed AnalyzerPluginSupportInfo_{A,B} to new name:
AnalyzerPluginSupportInfo.

analyzerbeats.cpp:
- Added translation unit for analyzerpluginsupport.
- Moved common functions from analyzer{beats,key} into
analyzerpluginsupport.
- Made closure object inline.
- Removed extra definition.

daschuer commented on src/analyzer/plugins/analyzerpluginsupport.h:
```diff
@@ -0,0 +1,19 @@
+#include "analyzer/analyzer.h"
+#include "analyzer/plugins/analyzerplugin.h"
+
+struct AnalyzerPluginSupportInfo {
```
I think this should be a class. We use struct for structs without member functions.

daschuer commented on src/analyzer/plugins/analyzerpluginsupport.h:
```diff
@@ -0,0 +1,19 @@
+#include "analyzer/plugins/analyzerplugin.h"
+
+struct AnalyzerPluginSupportInfo {
+    virtual QString matchAndSetPluginId(
```
virtual is used if the child class implements the function. I think this is not the case here, so we should remove the virtual keyword.

daschuer commented on src/analyzer/analyzerbeats.cpp:
```diff
@@ -136,7 +120,7 @@ bool AnalyzerBeats::shouldAnalyze(TrackPointer pTrack) const {
     QString pluginID = m_bpmSettings.getBeatPluginId();
     if (pluginID.isEmpty()) {
-        pluginID = defaultPlugin().id();
+        pluginID = defaultPlugin(availablePlugins()).id();
```
Idea: you can make availablePlugins() an abstract virtual function ( = 0 ) of AnalyzerPluginSupportInfo;
Than we get around passing availablePlugins(). 
Do we always use the id()? You may consider to return that directly.

daschuer commented on src/analyzer/analyzerkey.h:
```diff
@@ -6,18 +6,18 @@
 #include "util/memory.h"
 
-class AnalyzerKey : public Analyzer {
+class AnalyzerKey : protected AnalyzerPluginSupportInfo, public Analyzer {
```
protected inheritance is uncommon in C++ because in addition to the already protected functions it makes the public functions also protected. It looks like in this case you can inherit privately. Otherwise inherit public and make all functions protected.

daschuer reviewed (changes requested):
Thank you for this nice clean up PR.  I have added some comments. 
Feel free to rebase/squash this PR as you like. It is OK in  this case because of the limited scope.

tcoyvwac committed:
analyzer: abstracted to virtual mem-fn.

analyzerpluginsupport:
- availablePlugins(2), now virtual function template. (Currently name
blocked by previous API design).
- defaultPlugin() now returns id of plugin directly.
- matchAndSetPluginId() changed to accept availablePlugins(2) virtual
templated version.

analyzer: adjusted static member-function dependencies

analyzer{beats,key}.h:
Created defaultAvailable{Beats,Key}Plugins() for dialogs/
dlgpref{beats,key} translation units to (statically) call. This frees
API naming so AnalyzerPluginSupportInfo can internally use the function
name: availablePlugins.

tcoyvwac commented:
Hi @daschuer, thank you for the positive feedback!
_(BTW, when closer to the A-OK merging-point, I will rebase all commits to make history a bit better.)_

---
Notes:

* I totally agree with your points on inheritance, so I adjusted the inheritance-scoping to match what was recommended by B.Stroustrup in his FAQ.
  https://www.stroustrup.com/bs_faq2.html#abstract-class

* On making `availablePlugins()` a pure virtual function...
  *  It was discovered some other translation units elsewhere were using the statically-declared member function.
  * I had to adjust their translation units by moving / making the previous static member function its own (static) uniquely-name-defined free-function. 
  * I then adjusted any translation units which needed this free-function's data.
  * This should be in scope of this PR, however if too large of a scope, I can happily remove the (final) commit and make a small separate PR to do with this slight API change if you like.

daschuer commented on src/analyzer/plugins/analyzerpluginsupport.cpp:
```diff
@@ -0,0 +1,27 @@
+            return info.id() == pluginIdToMatch;
+        })
+                ? pluginIdToMatch // configured Plug-In available;
+                : defaultPlugin();
```
This conditional assignment is hard to read. How about using if/else here?

daschuer commented on src/analyzer/plugins/analyzerpluginsupport.h:
```diff
@@ -0,0 +1,19 @@
+    QString m_pluginId;
+};
+
+class AnalyzerPluginSupportInfo : protected AnalyzerPluginIdMetadata {
```
Can we just merge m_pluginId into AnalyzerPluginSupportInfo?

daschuer commented on src/analyzer/plugins/analyzerpluginsupport.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "analyzer/plugins/analyzerpluginsupport.h"
+
+QString AnalyzerPluginSupportInfo::matchAndSetPluginId(const QString& pluginIdToMatch) {
```
set functions normaly set a member variable which is not the case here. 
How about renaming this to "matchOrGetDefaultPluginId" or just "matchPlugInID" or ...

daschuer commented:
Thank you! This looks already nice. I have added some suggestions for improvement.

tcoyvwac committed:
analyzer: rename function name

analyzerpluginsupport:
- Renamed function to matchOrGetDefaultPluginId().
- Converted ternary operation to if/else statment.

daschuer commented:
Ok, go ahead. 
This LGTM, give me a hint if it is ready for merge.

tcoyvwac commented:
@daschuer 
Ok, the changesets / commits history has been redone to be much clearer. This PR is now complete if there is nothing else required. Please feel free to merge, ye? :smile:

daschuer commented on src/analyzer/analyzerbeats.cpp:
```diff
@@ -13,20 +13,16 @@
 
-// static
-QList<mixxx::AnalyzerPluginInfo> AnalyzerBeats::availablePlugins() {
+QList<mixxx::AnalyzerPluginInfo> defaultAvailableBeatsPlugins() {
```
This is now a global free function without a namespace. I think In this case it is best to just manually inline it into  
AnalyzerBeats::availablePlugins() and use that one instead.

tcoyvwac replied:
Good eye and agreed on removing the global free-function. Now made the function namespaced.

---

Looking at the code design, while refactoring to summarise a simple common plugins interface, it seems that (the terms) "`availablePlugins`" and "`defaultPlugins`" have a different use-purpose.

To check to see what plugins are available: there is the supposedly ["availablePlugins" function API.](#diff-4186bd7b4da09d30a81e5b021b8991aacb26f9e75310c37403199da822c506beL60-L70)

To use the appropriate initial default plugins for the {Beats,Key}Plugin, "[defaultPlugins" is the API to use. ](https://github.com/mixxxdj/mixxx/blob/6bf5a0a94bdd1f1152f19ba30b25e29d7731b6e1/src/analyzer/analyzerkey.cpp#L16-L24)

From observations, it seems planned future-wise, if there are more "plugins", the "defaultPlugins" {Beats,Key} can differ depending on the compile-time settings, but the API to always get the "availablePlugins" must be somewhat stable, to "perform" the same API requirements regardless of whichever (future) plugin.

---

In the end, if the main blocking concern was the global free-function, then the rest of the PR should be fine and ok to merge these changes to mainline.

daschuer reviewed (changes requested):
Thank you.

tcoyvwac committed:
analyzer: namespaced global free-function

Swiftb0y commented on src/analyzer/analyzerkey.h:
```diff
@@ -6,19 +6,19 @@
-class AnalyzerKey : public Analyzer {
+QList<mixxx::AnalyzerPluginInfo> defaultAvailableKeyPlugins();
+
+class AnalyzerKey : private AnalyzerPluginSupportInfo, public Analyzer {
```
using private inheritance as a means of code reuse seems very bad practice to me tbh.

tcoyvwac replied:
`AnalyzerPluginSupportInfo` is an interface. The simple interface design came about after seeing duplication of code between the (currently) two classes `Analyzer{Beats,Key}`. The only classes needing to use this interface and access any member-variables within, are `Analyzer{Beats,Key}`, _and no other external translation-units,_ that is why the interface `AnalyzerPluginSupportInfo` is declared `private` (for them).

The bonus of consolidating the common code into an interface-style is the code reduction (and natural beneficial code-reuse).

Some of this approach is explained in these articles:
* https://www.stroustrup.com/bs_faq2.html#abstract-class
* https://www.artima.com/articles/the-c-style-sweet-spot

I am happy to declare it something other than `private`, but as no other translation-unit is using the interface, it would seem slightly unorthodox.

Swiftb0y commented on src/analyzer/plugins/analyzerpluginsupport.cpp:
```diff
@@ -0,0 +1,29 @@
+            return defaultPlugin();
+        }
+    }
+    return m_pluginId;
```
I struggle to understand this reasoning. Why do we return `m_pluginId` when no plugins are available but above we return the `defaultPlugin`?

tcoyvwac replied:
[This is a translation of the previous code (and its specifications).](#diff-4186bd7b4da09d30a81e5b021b8991aacb26f9e75310c37403199da822c506beL60-L70)

As the function `matchOrGetDefaultPluginId` assigns its results to a variable after returning, according to the old code, the function should return `m_pluginId` value if the first predicate-check fails.

As for _the design spec & functionality of the old code,_ well... I don't know about that specification when it was created... but that would be outside of the scope of this consolidation PR.

Swiftb0y replied:
Mhmm. I don't like the fact that we're returning `m_pluginId`. Yes, its (almost) the logical equivalent of what its replacing, but it doesn't match the function name and it also assumes that the function is used only when assigning to `m_pluginId`. I mean almost because if we're super pedantic about it, it doesn't specify `m_pluginId` is returned and assigned, it means `m_pluginId` is not being touched. Depending on the type, this makes a difference (though it probably doesn't since `m_pluginId` is just a (likely empty) Qstring).

My biggest problem with this function is, that its not obvious that it will return `m_pluginId` instead of the default when there are no plugins available. That is not obvious from the function definition and the reason is even less obvious when looking at the implementation.

tcoyvwac replied:
I totally agree with your analysis.

>  I mean almost because if we're super pedantic about it, it doesn't specify m_pluginId is returned and assigned, it means m_pluginId is not being touched. Depending on the type, this makes a difference (though it probably doesn't since m_pluginId is just a (likely empty) Qstring).

From the original old code-section, it was unclear what was happening with `m_pluginId` / `m_pluginId` purpose. I wanted to make the section "testable", and extracted the code portion into its own function. This makes the code (and its odd requirements / specifications) clearer, however, your point stands. Quietly, I am happy that it has brought into question (the code's spec and) further code-design discussions.

Again, this is outside of the scope of the consolidation PR, but having duplicated code, which was semi-unclear in its purpose, I felt that this rewrite at least clears up the code's spec a bit more (for all). I wanted to push these adjusted changes to mainline (as good practice) and further work on the code sections on a later date.

tcoyvwac replied:
> ...it means `m_pluginId` is not being touched. 

(This comment made me realise that the function was not marked as `const`. This has now been corrected).

Swiftb0y replied:
Thank very much you for clarifying. Even though I'm grateful for your contribution, my opinion stated below still stands. So consider my review dismissed.

acolombier commented:
> Thank you.

@daschuer did you mean to approve this PR? (You previously did)
If it's fine by you, I think this could be merged

@tcoyvwac I'll set this PR as draft, but feel free to mark it ready if you get a chance to address the conflicts

ronso0 commented:
2 months without response
@daschuer hope it's okay to close this now.
_____
I'll close this abandoned PR now to clean up the list of open PRs.
If the author or maintainers think this should remain open please reopen.
If anyone thinks this is worth salvaging and is motivated to pick it up, resolve the conflicts and move it forward, feel free to open a new PR.
